kernel/src/engine/arrow_utils.rs

```diff
@@ -459,6 +459,52 @@ fn get_indices(
     ))
 }
 
+/// Contains information about a StructField
+/// # Fields:
+/// * `0` - The index of the struct field in its struct
+/// * `1` - A reference to the struct field's name
+/// * `2` - A reference to the struct field
+type FieldInfo<'k> = (usize, &'k String, &'k StructField);
+
+/// Constructs an iterator where each parquet Field in `fields` is matched
+/// with a a kernel `FieldInfo` reperesenting a StructField.
+///
+/// This returns an iterator of tuples. Each tuple contains the parquet field's ordinal
+/// and the matching `FieldInfo` if present.
+fn get_field_infos<'k, 'p>(
+    requested_schema: &'k StructType,
+    fields: &'p Fields,
+) -> impl Iterator<Item = (usize, &'p Arc<Field>, Option<FieldInfo<'k>>)> {
+    // Construct a map from the field id to its StructField
+    let field_id_to_name: HashMap<i64, &String> = requested_schema
+        .fields()
+        .filter_map(
+            |field| match field.get_config_value(&ColumnMetadataKey::ParquetFieldId) {
+                Some(MetadataValue::Number(fid)) => Some((*fid, field.name())),
+                _ => None,
+            },
+        )
+        .collect();
+
+    // Closure to map the parquet Field to the matching kernel FieldInfo if present
+    let parquet_to_field_info = move |field: &'p Field| -> Option<FieldInfo<'k>> {
+        // Get field id
+        let field_id_opt = field
+            .metadata()
+            .get(PARQUET_FIELD_ID_META_KEY)
+            .and_then(|x| x.parse::<i64>().ok());
+        let field_name = field_id_opt
+            .and_then(|field_id| field_id_to_name.get(&field_id).copied())
+            .unwrap_or_else(|| field.name());
+        requested_schema.fields.get_full(field_name)
+    };
+
+    fields
+        .iter()
+        .enumerate()
+        .map(move |(parquet_index, field)| (parquet_index, field, parquet_to_field_info(field)))
```

Comment:
nit: can you add a comment that we need to `move` `parquet_to_field_info` into the iterator so we can return it